File: firstNovel/CKSettingsViewController.m
//
//  CKSettingsViewController.m
//  firstNovel
//
//  Created by 张超 on 1/19/14.
//  Copyright (c) 2014 张超. All rights reserved.
//

#import "CKSettingsViewController.h"
#import "CKCommonUtility.h"
#import "BBADownloadManagerViewController.h"
#import "CKRootViewController.h"
#import "BBADownloadDataSource.h"
#import "CKRootViewController.h"

enum ESettingSection {
    TSettingSectionDM = 0,
    TSettingSectionCommon,
    TSettingSectionCount
};

enum ESettingSectionDM {
    TSettingDMRowDownloadManager = 0,
    TSettingDMRowCount
};

enum ESettingSectionCommon {
    TSettingCommonRowFeedBack = 0,
    TSettingCommonRowRate,
    TSettingCommonRowCount
};

@interface CKSettingsViewController ()

@end

@implementation CKSettingsViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self)
    {
        // Custom initialization
    }
    return self;
}

- (void)dealloc
{
    [_settingsTable release];
    [UMFeedback sharedInstance].delegate = nil;;
    
    [super dealloc];
}

- (void)loadView
{
    [super loadView];
    
    self.view.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"main_view_bg.png"]];
    
    _settingsTable = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(IOS_7_0))
    {
        _settingsTable.frame = CGRectMake(0.0f, STATUS_HEIGHT + NAVIGATIONBAR_HEIGHT, APPLICATION_FRAME_WIDTH, APPLICATION_FRAME_HEIGHT - TABBAR_HEIGHT - STATUS_HEIGHT - NAVIGATIONBAR_HEIGHT);
    }
    else
    {
        _settingsTable.frame = CGRectMake(0.0f, 0.0f, APPLICATION_FRAME_WIDTH, APPLICATION_FRAME_HEIGHT - TABBAR_HEIGHT  - NAVIGATIONBAR_HEIGHT);
    }
    _settingsTable.dataSource = self;
    _settingsTable.delegate = self;
    _settingsTable.allowsMultipleSelection = NO;
    _settingsTable.allowsSelectionDuringEditing = NO;
    _settingsTable.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
    _settingsTable.backgroundColor = [UIColor clearColor];
    _settingsTable.backgroundView = [[[UIView alloc] init] autorelease];
    [self.view addSubview:_settingsTable];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [[UMFeedback sharedInstance] setAppkey:UMENG_APPKEY delegate:self];
	// Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:nil] autorelease];
    //cell.backgroundColor = [UIColor clearColor];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    if (!SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(IOS_7_0))
    {
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    if (indexPath.section == TSettingSectionDM)
    {
        cell.textLabel.text = @"下载的小说";
        cell.imageView.image = [UIImage imageNamed:@"settings_download.png"];
        if ([[BBADownloadDataSource sharedInstance] totalCount] > 0)
        {
            cell.detailTextLabel.text = [NSString stringWithFormat:@"%d", [[BBADownloadDataSource sharedInstance] totalCount]];
        }
    }
    else if (indexPath.section == TSettingSectionCommon)
    {
        if (indexPath.row == TSettingCommonRowFeedBack)
        {
            cell.textLabel.text = @"给点建议";
            cell.imageView.image = [UIImage imageNamed:@"settings_feedback.png"];
        }
        else if (indexPath.row == TSettingCommonRowRate)
        {
            cell.textLabel.text = @"给应用评分";
            cell.imageView.image = [UIImage imageNamed:@"settings_rate.png"];
        }
    }
    cell.textLabel.font = [UIFont systemFontOfSize:15.0f];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section == TSettingSectionDM && indexPath.row == TSettingDMRowDownloadManager)
    {
        BBADownloadManagerViewController *downloadViewController = [[BBADownloadManagerViewController alloc] init];
        [[CKRootViewController sharedInstance].rootNaviViewController pushViewController:downloadViewController animated:YES];
        [downloadViewController release];
        
        [MobClick event:@"settingsDownload"];
    }
    else if (indexPath.section == TSettingSectionCommon && indexPath.row == TSettingCommonRowFeedBack)
    {
        [MobClick event:@"settingsFeedback"];
        [UMFeedback showFeedback:[CKRootViewController sharedInstance] withAppkey:UMENG_APPKEY];
    }
    else if (indexPath.section == TSettingSectionCommon && indexPath.row == TSettingCommonRowRate)
    {
        [MobClick event:@"settingsRate"];
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (section == TSettingSectionDM)
    {
        return TSettingDMRowCount;
    }
    else if (section == TSettingSectionCommon)
    {
        return TSettingCommonRowCount;
    }
    return 0;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return TSettingSectionCount;
}

- (void)getFinishedWithError: (NSError *)error
{
    NSLog(@"%@", error);
}

- (void)postFinishedWithError:(NSError *)error
{
    NSLog(@"%@", error);
}

@end

File: scripts/CKSettingsViewController.lua
waxClass{"CKSettingsViewController", protocols={"UITableViewDataSource", "UITableViewDelegate", "UMFeedbackDataDelegate"}}

function loadView(self)
    self.super:loadView()
    self:view():setBackgroundColor(UIColor:colorWithPatternImage(UIImage:imageNamed("main_view_bg.png")))
    self.settingsTable = UITableView:initWithFrame_style(CGRect(0.0, 0.0, 0.0, 0.0), UITableViewStyleGrouped)
    self.settingsTable:setFrame(CGRect(0.0, 0.0, 0.0, 0.0))
    local result = wax.bba.SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO("7.0")
    if result then
        self.settingsTable:setFrame(CGRect(0.0, STATUS_HEIGHT + NAVIGATIONBAR_HEIGHT, wax.bba.APPLICATION_FRAME_WIDTH(), wax.bba.APPLICATION_FRAME_HEIGHT() - TABBAR_HEIGHT - STATUS_HEIGHT - NAVIGATIONBAR_HEIGHT))
    else
        self.settingsTable:setFrame(CGRect(0.0, 0.0, wax.bba.APPLICATION_FRAME_WIDTH(), wax.bba.APPLICATION_FRAME_HEIGHT() - TABBAR_HEIGHT  - NAVIGATIONBAR_HEIGHT))
    end
    self.settingsTable:setDataSource(self)
    self.settingsTable:setDelegate(self)
    self.settingsTable:setAllowsMultipleSelection(false)
    self.settingsTable:setAllowsSelectionDuringEditing(false)
    self.settingsTable:setSeparatorStyle(UITableViewCellSeparatorStyleSingleLine)
    self.settingsTable:setBackgroundColor(UIColor:clearColor())
    self.settingsTable:setBackgroundView(UIView:init())
    self:View():addSubView(self.settingsTable)
end

function viewDidLoad(self)
    self.super:viewDidLoad()
    UMFeedback:sharedInstance():setAppkey_delegate("52d369b456240b8c500ef4c1", self)
end

function tableView_cellForRowAtIndexPath(self, tableView, indexPath)
    local cell = UITableViewCell:initWithStyle_reuseIdentifier(UITableViewCellStyleValue1, nil)
    cell:setBackgroundColor(UIColor:clearColor())
    cell:setAccessoryType(UITableViewCellAccessoryDisclosureIndicator)
    cell:setSelectionStyle(UITableViewCellSelectionStyleNone)
    if indexPath:row() == TSettingCommonRowFeedBack then
        cell:textLabel():setText("给点建议")
        cell:imageView():setImage(UIImage:imageNamed("settings_feedback.png"))
    elseif indexPath:row() == TSettingCommonRowRate then
        cell:textLabel():setText("给应用评分")
        cell:imageView():setImage(UIImage:imageNamed("settings_rate.png"))
    elseif indexPath:row() == TSettingCommonRowVersion then
        cell:textLabel():setText("版本")
        cell:imageView():setImage(UIImage:imageNamed("settings_rate.png"))
        cell:detailTextLabel():setText(wax.bba.XcodeAppVersion())
    end
    cell:textLabel():setFont(UIFont:systemFontOfSize(15.0))
    return cell
end

function tableView_didSelectRowAtIndexPath(self, tableView, indexPath)
    if indexPath:section() == TSettingSectionCommon and indexPath:row() == TSettingCommonRowFeedBack then
        MobClick:event("settingsFeedback")
        UMFeedback:showFeedback_withAppkey(CKRootViewController:sharedInstance(), "52d369b456240b8c500ef4c1")
    elseif indexPath:section() == TSettingSectionCommon and indexPath:row() == TSettingCommonRowRate then
        MobClick:event("settingsRate")
        CKCommonUtility:goRating()
    end
end

function tableView_numberOfRowsInSection(self, tableView, section)
    if section == TSettingSectionCommon then
        return TSettingCommonRowCount
    end
    return 0
end

function numberOfSectionsInTableView(self, tableView)
    return TSettingSectionCount
end




